app/src/main/java/com/example/gameworkflow/TimeMachineWorkflow.kt

```diff
@@ -0,0 +1,85 @@
+package com.example.gameworkflow
+
+import com.example.gameworkflow.TimeMachineWorkflow.State
+import com.squareup.workflow.*
+import com.squareup.workflow.WorkflowAction.Mutator
+
+/**
+ * Workflow that records its child's renderings and allows you to travel back in time.
+ */
+class TimeMachineWorkflow(
+    private val child: Workflow<Unit, Nothing, Any>,
+    private val shakeWorker: ShakeWorker
+) : StatefulWorkflow<Unit, State, Nothing, Any>() {
+
+    data class State(
+        val replayPoint: Int = PRESENT,
+        val history: MutableList<Any> = mutableListOf()
+    )
+
+    /**
+     * A screen that shows a previous rendering of the child, with a seek bar for scrubbing time.
+     *
+     * @param screen The child rendering to display.
+     * @param historyEnd The last valid position in the history.
+     * @param historyPosition The position in the history that [screen] is from.
+     * @param onSeek Event handler to call to scrub the history.
+     */
+    data class TimeTravelScreen(
+        val screen: Any,
+        val historyEnd: Int,
+        val historyPosition: Int,
+        val onSeek: (position: Int) -> Unit
+    )
+
+    private val shakeAction = WorkflowAction<State, Nothing> {
```

Comment:
s/shakeAction/handleShaking/

Active name like this will make the `runningWorker` block below more readable.

Reply:
Went with `onShake` to be consistent with `onSeek`.